### PR TITLE
Value cache

### LINKED ISSUE #20006: Add kd-tree based value cache
## Reason
Material models with internal solves could benefit from caching values index by _R^n_ vectors. E.g. solving the phase concentrations in a Kim-Kim-Suzuki phase field model can take a lot of iteration. Similarly Gibbs Energy minimization in Yellowjacket can be expensive. The kd-tree based cache could be a way to provide good initial guesses to improve convergence.

## Design
Add templated cache helper class using `KDTreeSingleIndexDynamicAdaptor` (requires libmes/nanoflann update).

## Impact
Improved performance (remains to be benchmarked)

### PR DESCRIPTION
**Pull request by dschwen**
Closes #20006

It looks like the libmesh update hasn't made it to devel yet...

### CODE REVIEW

File: framework/include/utils/ValueCache.h [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+
+template <typename T>
+ValueCache<T>::ValueCache(std::size_t in_dim, std::size_t max_leaf_size)
+  : _point_cloud(in_dim),
+    _kd_tree(nullptr),
+    _in_dim(in_dim),
+    _max_leaf_size(max_leaf_size),
+    _max_subtrees(100)
+{
+  mooseAssert(_kd_tree != nullptr, "KDTree was not properly initialized.");
```
Comment (loganharbour) [inline framework/include/utils/ValueCache.h:76]: This does not make sense to me (the assertion)

**Comment by cticenhour** [comment]
@dschwen the libmesh update just made it to devel, FYI.

**Review by loganharbour — APPROVED** [review]
Have a place to use this somewhere?

**Comment by dschwen** [comment]
I have two pending applications, one in Yellowjacket, and one phase field related.

**Comment by loganharbour** [comment]
Sounds good. I'm fine with it going in whenever you want, then.

**Pull request merged**